### DISCUSSION TITLE
How do I build a C++ program with tree-sitter and tree-sitter-cpp on Windows

### DISCUSSION THREAD

**Discussion by JonnyKelso** [tree-sitter/tree-sitter#2141] [Q&A] [answered]
Hello, I am trying to write a new application that uses tree-sitter to parse c++.
I've been trying to follow the instructions on [https://tree-sitter.github.io/tree-sitter/using-parsers](https://tree-sitter.github.io/tree-sitter/using-parsers), and I managed to build the little test program `test-json-parser`, and I can run that just fine.

Now I want to try and set up a new test program for C++ parsing, so I cloned `tree-sitter/tree-sitter-cpp` and modified the test-json-parser code to set the `TSLanguage` to `tree_sitter_cpp()`

When I build it, I get numerous linker errors. I am building with the clang command:
`clang -I tree-sitter/lib/include test-cpp-parser.c tree-sitter-cpp/src/parser.c tree-sitter/bin/tree-sitter.a -o test-cpp-parser.exe`
I get the following linker errors:
```
tree-sitter-cpp/src/parser.c:9:9: warning: '#pragma optimize' is not supported [-Wignored-pragma-optimize]
#pragma optimize("", off)
        ^
1 warning generated.
   Creating library test-cpp-parser.lib and object test-cpp-parser.exp
parser-984318.o : error LNK2001: unresolved external symbol tree_sitter_cpp_external_scanner_create
parser-984318.o : error LNK2001: unresolved external symbol tree_sitter_cpp_external_scanner_destroy
parser-984318.o : error LNK2001: unresolved external symbol tree_sitter_cpp_external_scanner_scan
parser-984318.o : error LNK2001: unresolved external symbol tree_sitter_cpp_external_scanner_serialize
parser-984318.o : error LNK2001: unresolved external symbol tree_sitter_cpp_external_scanner_deserialize
test-cpp-parser.exe : fatal error LNK1120: 5 unresolved externals
clang: error: linker command failed with exit code 1120 (use -v to see invocation)
```

It seems like it can't find the tree-sitter-cpp functions, but I'm using the same clang command line as with the json test example, but modified to compile `tree-sitter-cpp/src/parser.c` instead.
I'm building it as a C program, but ultimately I'd like to use a C++ compiler and write my program in C++ with clang++

**Comment by sogaiu** [comment, marked as answer]
I think that grammar has an [external scanner](https://github.com/sogaiu/ts-questions/blob/master/questions/are-external-scanners-common/README.md) portion so [that source](https://github.com/tree-sitter/tree-sitter-cpp/blob/master/src/scanner.cc) needs to be accounted for when compiling.

The following is part of a shell script that accounts for that kind of situation: https://github.com/casouri/tree-sitter-module/blob/d3ed3beb47dcc7b73098f636af737f489fb5c93c/build.sh#L108-L127

**Comment by JonnyKelso** [comment]
Thanks, that seems ot have worked so far... for future searchers, here is what worked for me:
I used this to build the tree-sitter-cpp scanner object file:
```
cd tree-sitter-cpp
clang -c -I../tree-sitter/lib/include src/scanner.cc -o scanner.o
```
then this to build tree-sitter itself:
```
cd tree-sitter
clang  -c -O0 lib/src/lib.c -Ilib/src -Ilib/include -o tree-sitter.a
```
then this to build the example (but I modified the exmaple so it was parsing C++ instead of some json):
```
cd myapp // which contains tree-sitter and tree-sitter-cpp folders too
clang -I tree-sitter/lib/include test-cpp-parser.c tree-sitter-cpp/src/parser.c tree-sitter/bin/tree-sitter.a tree-sitter-cpp/scanner.o -o test-cpp-parser.exe
```

**Reply by sogaiu** [reply]
Thanks for sharing what worked in your situation :)
